[CORE] Adding support for insertion of soft-tuned prompts

SwapnilDreams100 opened the pull request:
This PR adds support for inserting soft-tuned prompts into the input embeddings (trained using [PEFT](https://huggingface.co/docs/peft/main/en/conceptual_guides/prompting)).

This functionality is required by the IBM team.

Summary of Changes:
- New `prompt_adapter` folder similar to the `lora` folder to create a LRU cache management system for multiple prompt adapters
- New `prompt_adapter_config` engine parameter - for easy extension to more sophisticated prompt-tuning techniques in the future
- New `prompt_adapter_request` parameter added to the generate functionality of the LLM_Engine
- Current support for several models like `bloom, llama, mistral`, easily extensible for others
- Simple test demonstrating that prompt adapters work for `bloom`
- Some parameter documentation still pending, will complete post successful code review

Yard1 commented on vllm/model_executor/models/baichuan.py:
```diff
@@ -268,6 +268,13 @@ def forward(
         attn_metadata: AttentionMetadata,
     ) -> torch.Tensor:
         hidden_states = self.embed_tokens(input_ids)
+        if hasattr(self, 'prefix_encoder'):
```
can we set `self.prefix_encoder` as None by default and then turn this check into an `is not None`?

also, if this logic is going to be the same for every model, can we separate it out into a function and reuse?

SwapnilDreams100 replied:
1. Sure that sounds good.
2. Yes we can do that, we can abstract this out into a function which takes in the embeddings and positions and spits out the new hidden states

Yard1 commented on vllm/model_executor/models/baichuan.py:
```diff
@@ -268,6 +268,13 @@ def forward(
+            indices = (positions < soft_prompt.shape[0])
+            # Replace position_emb into hidden_states at the found indices
+            hidden_states[indices] = soft_prompt[positions[indices]]
+        else:
+            pass
```
unnecessary else

Yard1 commented on vllm/prompt_adapter/models.py:
```diff
@@ -0,0 +1,231 @@
+    return _GLOBAL_PROMPT_ADAPTER_ID
+
+
+class PromptAdapterModel(object):
```
I think all of this makes a lot of sense, but it makes me wonder if we can promote code reuse between Prompt and Lora managers by eg. making a generic version of those two and then subclass/compose into specialized versions. Right now a lot of code seems duplicated.

SwapnilDreams100 replied:
Yes I agree, we can do that

Yard1 reviewed (commented):
Some high level comments - if this is not time sensitive, I think it would be good if we could come up with some more generic APIs to promote code reuse between LoRA and prompt adapter

Commit:
soft prompt support

Commit:
Run yapf and ruff

Commit:
Multimodal fix

Commit:
correctness update

Commit:
formatting

Commit:
formatting

Commit:
reverting to hasattr

Commit:
adapter commons fix

Commit:
minor fixes

Commit:
formatting

Commit:
reset_adapter

Commit:
bugfix

Commit:
reset_adapter fix

Commit:
peft dependencies

Commit:
fixing llava bug

Commit:
typing fix

Commit:
async engine update

Commit:
batchwise processing

Commit:
formatting

Commit:
formatting yapf

Commit:
formatting again

Commit:
enable_adapter paramter

Commit:
formatting

Commit:
adding test

Commit:
adding test

SwapnilDreams100 commented:
Hey @Yard1, 
I have addressed your comments on this soft prompt tuning PR. Some updates:
- New `adapter_commons` folder with all the common code between `LoRA` and `Prompt Adapters` abstracted
- Reduced redundancy of having to add the adapter addition code in all model classes by extending the `VocalParallelEmbedding` layer
- New parameter `enable_prompt_adapter` for more consistency w `LoRA`
- Added testing for multi-adapter inference

After your review, happy to add more extensive tests + docs.

Commit:
test case update

Commit:
formatting

Yard1 commented:
@SwapnilDreams100 thanks! let me take a look

SwapnilDreams100 commented:
Hi @Yard1 just a friendly reminder to review this PR when you get a chance, thanks!

g-eoj commented:
Hi @SwapnilDreams100, I have an initial implementation of adapter support for the OpenAI entry points based on https://github.com/SwapnilDreams100/vllm/tree/main. Would you be open to me contributing to your PR?

Commit:
resetting

Commit:
formatting

Commit:
formatting

Commit:
formatting

Yard1 commented:
@SwapnilDreams100 Sorry for the delay here, I should be able to review this before the end of week.

SwapnilDreams100 commented:
@Yard1 no worries, thank you!
Yes it should work, you can provide both PromptAdapterRequest and LoRaRequest parameters. I just tested a tiny example of this, happy to add a test on this.

Yard1 commented:
got it!

Commit:
Fix async engine

Commit:
Initial implementation of openai entrypoint

Assumes the interface for prompt adapters and lora modules remains
completely separate.

SwapnilDreams100 committed:
Merge branch 'main' into main

Commit:
Fixes

Yard1 commented on vllm/lora/models.py:
```diff
@@ -469,20 +472,24 @@ def activate_lora(
-            self._active_loras.pop(lora_id)
-            return True
-        return False
+    @property
+    def _deactivate_adapter(self):
+        return self._deactivate_lora
+
+    @property
+    def deactivate_lora(self):
+        return self.deactivate_adapter
```
this is a bit of an odd pattern (property returning a method)

SwapnilDreams100 replied:
@Yard1 The reason for this pattern is that I want to keep the original LoRA functions untouched while also abstracting the common functions in the base adapter classes.
So for instance, the base adapter class has methods like `deactivate_adapter` which is common between both lora and prompt-adapter codebases so the functionality has been written there. But in order to keep the LoRA function `deactivate_lora` I am routing it to the base function.
Any other way we should solve this?

Yard1 replied:
I think it's just going to be better to do normal methods and call `super().`

Yard1 commented on vllm/prompt_adapter/layers.py:
```diff
@@ -0,0 +1,56 @@
+        self,
+        base_indices: List[int],
+    ):
+        self.indices = base_indices
```
have a copy that's a GPU tensor here
```suggestion
        self.indices_gpu = torch.tensor(base_indices, device="cuda")
```

SwapnilDreams100 replied:
sounds good

Yard1 commented on vllm/prompt_adapter/layers.py:
```diff
@@ -0,0 +1,56 @@
+        for idx in unique_indices:
+            if idx != 0:
+                pa_idx = self.embedding_tensors[idx].prompt_embedding
+                mask = (self.indices == idx)
```
```suggestion
                mask = (self.indices_gpu == idx)
```
try to not have any CPU<->GPU syncing in forward pass. let's also make sure cuda graphs work.

SwapnilDreams100 replied:
sounds good

Yard1 commented on vllm/prompt_adapter/layers.py:
```diff
@@ -0,0 +1,56 @@
+                    n_adapters = sum(mask) // pa_idx.shape[0]
+                    hidden_states[mask] = pa_idx.repeat(n_adapters, 1)
+                except Exception:
+                    pass
```
it seems weird to just suppress the exception. what's the desired behavior here? what exceptions are we expecting?

SwapnilDreams100 replied:
we do not expect any exceptions here, this slipped through from when I was doing my own testing, cleaned

Yard1 commented on vllm/prompt_adapter/layers.py:
```diff
@@ -0,0 +1,56 @@
+    pass
+
+
+class VocabParallelEmbeddingWithPromptAdapter(nn.Module):
```
Looking at this code it makes me thing it will clash with LoRA adapters (either the LoRA or this will not be ran at the same time if both are specified).

SwapnilDreams100 replied:
It doesn't seem to do this, since it extends the lora class which happens before this.
I have added a new test which tests having both lora and pa specified in the same request.

Yard1 commented on vllm/prompt_adapter/models.py:
```diff
@@ -0,0 +1,279 @@
+                 num_virtual_tokens=None,
+                 prompt_embedding=None) -> None:
+        self.id = prompt_adapter_id
+        self.kv_cache = None
```
where is that used?

SwapnilDreams100 replied:
it isn't we can remove it

Yard1 commented on vllm/prompt_adapter/models.py:
```diff
@@ -0,0 +1,279 @@
+                              torch_device='cuda') -> "PromptAdapterModel":
+        adapters_weights = load_peft_weights(adapter_model_and_path,
+                                             torch_device)
+        prompt_embedding = adapters_weights["prompt_embeddings"].half()
```
do we always want to cast to float16? this should probably be the actual model dtype.

SwapnilDreams100 replied:
sounds good

SwapnilDreams100 replied:
this has been changed to use the model dtype

Yard1 commented on vllm/prompt_adapter/models.py:
```diff
@@ -0,0 +1,279 @@
+            raise ValueError("No free prompt_adapter slots")
+        index, _ = first_free_slot
+        self._active_adapters[prompt_adapter_id] = None
+        prompt_adapter_model = \
```
avoid \ formatting, prefer parenthesis

SwapnilDreams100 replied:
sounds good

Yard1 commented on vllm/prompt_adapter/request.py:
```diff
@@ -0,0 +1,28 @@
+    prompt_adapter_id: int
+    prompt_adapter_local_path: str
+    prompt_adapter_num_virtual_tokens: int
+    __hash__ = AdapterRequest.__hash__
```
```suggestion
    def __hash__(self):
        return super().__hash__()
```

SwapnilDreams100 replied:
sounds good

Yard1 commented on vllm/worker/model_runner.py:
```diff
@@ -449,11 +471,25 @@ def _prepare_model_input(
                         multi_modal_kwargs_list[k].append(v)
 
+                if prompt_adapter_id > 0:
+                    prompt_adapter_requests.add(
+                        seq_group_metadata.prompt_adapter_request)
+
+                num_tokens = seq_group_metadata.\
+                                        prompt_adapter_num_virtual_tokens
+                pm = [prompt_adapter_id
+                      ] * num_tokens + [0] * (query_len - num_tokens)
+                prompt_adapter_index_mapping += pm
+                prompt_adapter_prompt_mapping.extend(
+                    [prompt_adapter_id] *
+                    (query_len if seq_group_metadata.sampling_params
+                     and seq_group_metadata.sampling_params.prompt_logprobs
+                     else 1))
```
let's put this in a separate method for cleanliness

SwapnilDreams100 replied:
sounds good

SwapnilDreams100 replied:
On this feedback point, it didn't feel intuitive to put just this code in a separate function, but keep the lora code just above this(which looks almost similar) inline. Also this code extends the variables in code, doing that in another function would be pretty wieldy. What do you think?

Yard1 replied:
ok let's keep it in here for now

g-eoj commented:
@SwapnilDreams100 openai entrypoint PR (https://github.com/SwapnilDreams100/vllm/pull/2) is ready for an initial review.

@SwapnilDreams100 @Yard1 if prompt adapters and lora modules can be used at the same time, should the model list returned by the api server list every possible adapter/module combination? If not, how should the user go about selecting adapter + model combos?

Thanks!

Rallio67 commented:
@SwapnilDreams100 @Yard1 I am very interested in this pull request working successfully. I have been testing the fork from Swapnil installing from source and while the tester code does seem to work with the bloomz-560m twitter classifier example, I can verify that it does not work with the modern implementations of PEFT and newer models like llama3. I have trained softprompts myself with llama3-8B and llama3-70B and verified the outputs using PEFT and huggingface transformers and they are working as expected, when I run the same softprompt using the configuration in the bloomz example the code is executing and there is some difference between the +/- prompt adapter, the prompt is clearly not be applied correctly and the outputs are pretty random. Please let me know if I can help more. I hope this can get merged soon, but important to verify it works with the models people are using today (llama, qwen, mistral, gemma, nemotron, etc).

SwapnilDreams100 commented on format.sh:
```diff
@@ -26,7 +26,7 @@ RUFF_VERSION=$(ruff --version | awk '{print $2}')
 CODESPELL_VERSION=$(codespell --version)
 ISORT_VERSION=$(isort --vn)
-CLANGFORMAT_VERSION=$(clang-format --version | awk '{print $3}')
+# CLANGFORMAT_VERSION=$(clang-format --version | awk '{print $3}')
```
sounds good

SwapnilDreams100 commented on vllm/adapter_commons/models.py:
```diff
@@ -0,0 +1,108 @@
+T = TypeVar('T')
+
+
+class AdapterLRUCache(LRUCache[T]):
```
this is meant to be used as a common base class for both prompt adapters and lora. It completely overlaps between the 2

SwapnilDreams100 commented on vllm/adapter_commons/models.py:
```diff
@@ -0,0 +1,108 @@
+        return super()._on_remove(key, value)
+
+
+class AdapterModelManager:
```
sounds good

SwapnilDreams100 commented on vllm/adapter_commons/models.py:
```diff
@@ -0,0 +1,108 @@
+        ...
+
+    def _deactivate_adapter(self, adapter_id: int):
+        raise NotImplementedError("Subclasses must implement this method.")
```
sounds good

SwapnilDreams100 commented on vllm/adapter_commons/models.py:
```diff
@@ -0,0 +1,108 @@
+    def add_adapter(self, adapter: Any) -> bool:
+        if adapter.id not in self._registered_adapters:
+            if len(self._registered_adapters) >= self.capacity:
+                raise RuntimeError("No free " + self.adapter_type + " slots.")
```
sounds good

SwapnilDreams100 commented on vllm/adapter_commons/models.py:
```diff
@@ -0,0 +1,108 @@
+            self._set_adapter_mapping(mapping)
+        self._last_mapping = mapping
+
+    def _set_adapter_mapping(self, mapping: Any) -> None:
```
@Yard1  the reason I implemented this private method is because `set_adapter_mapping` uses this internal method which needs to be implemented on the base adapter class level. 
If I only set this method in the child classes like _set_lora_mapping and _set_prompt_adapter mapping, then I won't be able abstract the common code in `set_adapter_mapping`.
Is there another way to do this?

Yard1 replied:
We may need to just redesign the API (which is fine).

SwapnilDreams100 commented on vllm/core/scheduler.py:
```diff
@@ -139,6 +140,10 @@ def __post_init__(self):
 
+        self.num_prompt_adapters: int = len(self.prompt_adapter_requests)
+        if self.num_prompt_adapters > 0:
+            self._sort_by_prompt_adapter_ids()
```
got it, I can remove this

SwapnilDreams100 commented:
1. @g-eoj thanks for the OpenAPI PR! Will look at it once I finish the refactor requested by Antoni.

SwapnilDreams100 commented:
@Rallio67 thanks for testing the PR out. 
I have tested it with bloom, llama-2 and gptbigcode and it seemed to work for me. 
Can you please share the lama-3 adapter you are using which is not working?

Interesting detail on the newer JSON and naming conventions, I had wrote this a while back so it may have changed, let me check this.

Commit:
async changes

Rallio67 commented:
@SwapnilDreams100 I have created a softprompt and some tester code that you can use to verify the method is working with llama-3 compatible models. I don't know what the issue might be, but llama-3 does use GQA instead of MHA for the attention for all model sizes. I can verify that the softprompt (uploaded as a .zip) and my python code using huggingface transformers and PEFT is working and generating the correct output. If you have your repo verified on your end as running the code equivalently I can do another verification from a fresh source install.

https://github.com/Rallio67/Softprompts-vllm

Thank you for working on this!

SwapnilDreams100 committed:
Merge branch 'main' into main

Commit:
formattign

Commit:
formatting

Rallio67 commented:
> Hey @Rallio67, the prompt works but there is a change pending which would cast the prompt to the model's dtype to make it work out of the box. Currently the prompts are casted to float16, which is how my own prompts were trained. So if you set dtype to float16 in the vLLM call: like so `LLM(model="unsloth/llama-3-8b", enable_prompt_adapter=True, dtype= 'half')` in the current version of the PR the prompt will work. The dtype-agnostic prompt change will be made in the next day or so, and then this should not be necessary.

I will test it out with that parameter when I get home in a few hours. If you were taking the bfloat16 values (which is format the prompt was saved in I think), then treating it directly as float16 would explain why the outputs were quite messed up. I will test it out and report back. I think vllm supports bfloat16 and float16 so hopefully the code will be able to handle using whatever data type the user works with.

Rallio67 commented:
I ran the bloom test code with llama-3-8b and my tester softprompt and can verify that the code works when:
```
llm = vllm.LLM(MODEL_PATH, enable_prompt_adapter=True, dtype = "half")
```

And the number of virtual tokens is changed to 20 from 8 (my adapter uses 20 virtual tokens). I think it will be good if you are able to have the user specify the precision, however for now the fp16 should be fine for most use cases so long as people are aware of this limitation. Bfloat16 is used pretty commonly also.

njhill commented:
float16 is generally preferred for inference when possible - the better precision means there's less output variability caused by different batching.

Commit:
adding dtype flexibility + pa lora refactor

SwapnilDreams100 commented on requirements-common.txt:
```diff
@@ -20,3 +20,4 @@ lm-format-enforcer == 0.10.1
 outlines >= 0.0.43 # Requires torch >= 2.1.0
 typing_extensions
 filelock >= 3.10.4 # filelock starts to support `mode` argument from 3.10.4
+peft
```
where would we add this instead?

Yard1 replied:
we should fail during initialization if prompt embeddings are enabled in engine and peft is not present

Commit:
formatting

Commit:
formatting

Commit:
xpu compatibility

Commit:
xpu compatibility

Commit:
xpu compatibility

Commit:
xpu compatibility

SwapnilDreams100 committed:
Merge branch 'main' into main

Commit:
formatting

Commit:
formatting + updating tests

Commit:
test changes

SwapnilDreams100 commented:
Hey @Yard1 I have addressed most of your comments in the latest version. A couple of points need some more discussion, pl clarify in their topics. Thanks!

- Automatic Dtype casting to model dtype instead of float16 casting
- Added test for both LoRA and PA used together in the same request - they can be used simultaneously, and cuda graphs work
- Changes to the `adapter_commons`, `lora` and `prompt_adapter` APIs to be cleaner. The `lora` functions needed to be renamed, but it only impacts the model_runner class and one lora test file.

Yard1 commented on vllm/prompt_adapter/layers.py:
```diff
@@ -0,0 +1,56 @@
+        self.indices_gpu = torch.tensor(base_indices, device="cuda")
+        self.flag = torch.sum(self.indices_gpu) > 0
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        hidden_states = self.base_layer(x)
+        if self.flag:
+            unique_indices = torch.unique(self.indices_gpu)
+            for idx in unique_indices:
+                if idx != 0:
+                    pa_idx = self.embedding_tensors[
+                        idx.item()].prompt_embedding
+                    mask = (self.indices_gpu == idx)
+                    n_adapters = sum(mask) // pa_idx.shape[0]
+                    hidden_states[mask] = pa_idx.repeat(n_adapters, 1)
+        return hidden_states
```
this code forces a CPU<->GPU sync, I am pretty sure it's not going to work with cuda graphs? can we prepare all inputs on GPU so that we do not have to do this? `embedding_tensors` should just be a preallocated tensor, similarly to how LoRA works.

Yard1 commented on vllm/adapter_commons/models.py:
```diff
@@ -0,0 +1,110 @@
+    def _deactivate_adapter(self, adapter_id: int):
+        raise NotImplementedError("Subclasses must implement this method.")
+
+    def deactivate_adapter(self, adapter_id: int) -> bool:
+        if adapter_id in self._active_adapters:
+            self._deactivate_adapter(adapter_id)
+            self._active_adapters.pop(adapter_id)
+            return True
+        return False
+
```
let's just make the public APIs abstract and push the implementation to subclasses

SwapnilDreams100 replied:
The code here is common between both lora and prompt adapter, do we still reimplement this in both subclasses?

SwapnilDreams100 replied:
Hey @Yard1 the code here is common between both lora and prompt adapter, do we still reimplement this in both subclasses?

Yard1 replied:
let's use util functions for code reuse

SwapnilDreams100 replied:
these functions mainly call other functions and internal objects, so won't be able to reuse some of them

SwapnilDreams100 replied:
actually scratch that, we pass the references to the class objects and they should be updated as well

Yard1 commented on vllm/adapter_commons/request.py:
```diff
@@ -0,0 +1,19 @@
+    """
+    Base class for adapter requests.
+    """
+
```
add abstract property for `adapter_id`

Yard1 commented on vllm/adapter_commons/worker_manager.py:
```diff
@@ -0,0 +1,63 @@
+        self._adapter_manager.set_adapter_mapping(mapping)
+
+    @abstractmethod
+    def _load_adapter(self, request: Any) -> Any:
```
likewise, let's make public methods abstract instead of private ones and push the implementation to subclasses (we can reuse code through functions)

Yard1 commented on tests/prompt_adapter/test_pa_lora.py:
```diff
@@ -0,0 +1,60 @@
+    return results
+
+
+def test_lora_prompt_adapter():
```
can we add a test where we have a dummy prompt embedding (which should have no effect on the output) and ensure that lora with dummy prompt embedding gives the same output as lora by itself?

Yard1 reviewed (changes requested):
Thanks, this is on a good track

Commit:
cpu-gpu sync changes + adapter abstract changes

Commit:
formatting

SwapnilDreams100 committed:
Merge branch 'main' into main

Commit:
rebase

Commit:
peft fix

Commit:
minor fix

Commit:
formatting

SwapnilDreams100 commented:
Hey @Yard1, thank you for your feedback! Pl review this version with your comments addressed:
Highlights:
- Converted all `adapter_commons` class methods to abstract, pushed common functions to `utils`
- Embedding tensor now preallocated. Added a new config param `max_prompt_adapter_tokens` about the max adapter size used to allocate that tensor
- `test_pa_lora` updated, now compares 2 lora requests, one with adapter and one w/o adapter

Yard1 commented on vllm/prompt_adapter/layers.py:
```diff
@@ -19,37 +20,58 @@ class VocabParallelEmbeddingWithPromptAdapter(nn.Module):
-                if idx != 0:
-                    pa_idx = self.embedding_tensors[
-                        idx.item()].prompt_embedding
+                if idx != -1:
```
Can you explain to me the general logic of applying those adapters, especially when can we expect the idx to be -1 or non-unique?

Yard1 replied:
Yeah we need to get to a point where forward pass has no:
* CPU<->GPU synchronization (no item calls, indexing with python lists etc.)
* No python control flow

maybe we can write a simple triton kernel for this? the logic seems quite straightforward...

SwapnilDreams100 replied:
Hey @Yard1, I have changed the logic to remove the loops and item calls, now it is only limited to one if statement called torch.numel which I think is CUDA graph safe.
Don't have a ton of experience in Triton, would need some time to figure that out if this isn't acceptable

SwapnilDreams100 replied:
Hey @Yard1, I have simplified the logic to remove the loops and item calls, now it is only limited to one if statement called torch.numel which I think is CUDA graph safe.
Don't have a ton of experience in Triton, would need some time to figure that out if this isn't acceptable

Yard1 replied:
We have to have absolutely 0 if statements :(

Let me see if I can write a quick triton kernel for this.

njhill replied:
@Yard1 do you mean this is needed for CUDA graphs to work? or just that it would result in a graph break which is sub-optimal?

SwapnilDreams100 replied:
@Yard1 I see `if full_output.ndim == 3:` in the forward of `VocabParallelEmbeddingWithLoRA`, is that CUDA graph safe? We can replace the torch.numel with this if this is acceptable.

Yard1 replied:
@njhill we can't have any cuda graph breaks since we capture the entire forward pass. this will mean the logic will be incorrect.

@SwapnilDreams100 any CPU control flow is not going to work (what will happen is that the cuda graph will be recorded going down one branch, and then when you replay that control flow logic will not be present in there, so you won't be able to dynamically change anything during runtime)

Yard1 reviewed (commented):
Great! This is looking really clean now API-wise. I am still a bit concerned about the logic to actually apply the adapters. Any python control flow logic (loops, ifs) will not be captured by the cuda graph. Can you explain to me how it's supposed to work?

prashantgupta24 commented on vllm/engine/async_llm_engine.py:
```diff
@@ -618,6 +632,7 @@ async def generate(
         request_id: str,
         lora_request: Optional[LoRARequest] = None,
         trace_headers: Optional[Dict[str, str]] = None,
+        prompt_adapter_request: Optional[PromptAdapterRequest] = None
```
Can we also add it to the function args docs below?

Commit:
forward update

Commit:
formatting

SwapnilDreams100 committed:
Merge branch 'main' into main

Commit:
formatting

Commit:
spec decode fix

SwapnilDreams100 committed:
Merge branch 'main' into main

Commit:
formatting

SwapnilDreams100 committed:
Merge branch 'main' into main

Commit:
async executor

Commit:
formatting

Commit:
formatting

Commit:
formatting

Commit:
Merge branch 'main' into openai-entrypoint

SwapnilDreams100 committed:
Merge branch 'main' into main

Commit:
formatting

g-eoj commented:
If `max_prompt_adapter_tokens` is less than `num_virtual_tokens`, there is an error like:
```
RuntimeError: The expanded size of the tensor (10) must match the existing size (128) at non-singleton dimension 0.
````

This message isn't intuitive but likely to be hit when using prompt adapters. 

- can the case of `max_prompt_adapter_tokens` is less than `num_virtual_tokens` be caught and a descriptive error message be shown?
- should `max_prompt_adapter_tokens` have  a default value? The user is more likely to set it correctly if they are forced to set it every time prompt adapters are enabled.

Commit:
max_prompt_adapter_token defaults + error messages

g-eoj commented:
@SwapnilDreams100 I'm hitting this error now with an adapter setup that worked before the refactor.

```
ERROR 07-03 19:42:16 async_llm_engine.py:54] Engine background task failed
ERROR 07-03 19:42:16 async_llm_engine.py:54] Traceback (most recent call last):
ERROR 07-03 19:42:16 async_llm_engine.py:54]   File "/usr/local/lib/python3.10/dist-packages/vllm/engine/async_llm_engine.py", line 44, in _log_task_completion
ERROR 07-03 19:42:16 async_llm_engine.py:54]     return_value = task.result()
ERROR 07-03 19:42:16 async_llm_engine.py:54]   File "/usr/local/lib/python3.10/dist-packages/vllm/engine/async_llm_engine.py", line 601, in run_engine_loop
ERROR 07-03 19:42:16 async_llm_engine.py:54]     result = task.result()
ERROR 07-03 19:42:16 async_llm_engine.py:54]   File "/usr/local/lib/python3.10/dist-packages/vllm/engine/async_llm_engine.py", line 546, in engine_step
ERROR 07-03 19:42:16 async_llm_engine.py:54]     request_outputs = await self.engine.step_async(virtual_engine)
ERROR 07-03 19:42:16 async_llm_engine.py:54]   File "/usr/local/lib/python3.10/dist-packages/vllm/engine/async_llm_engine.py", line 240, in step_async
ERROR 07-03 19:42:16 async_llm_engine.py:54]     output = await self.model_executor.execute_model_async(
ERROR 07-03 19:42:16 async_llm_engine.py:54]   File "/usr/local/lib/python3.10/dist-packages/vllm/executor/distributed_gpu_executor.py", line 173, in execute_model_async
ERROR 07-03 19:42:16 async_llm_engine.py:54]     return await self._driver_execute_model_async(execute_model_req)
ERROR 07-03 19:42:16 async_llm_engine.py:54]   File "/usr/local/lib/python3.10/dist-packages/vllm/executor/ray_gpu_executor.py", line 367, in _driver_execute_model_async
ERROR 07-03 19:42:16 async_llm_engine.py:54]     results = await asyncio.gather(*tasks)
ERROR 07-03 19:42:16 async_llm_engine.py:54]   File "/usr/local/lib/python3.10/dist-packages/vllm/executor/ray_gpu_executor.py", line 352, in _run_task_with_lock
ERROR 07-03 19:42:16 async_llm_engine.py:54]     return await task(*args, **kwargs)
ERROR 07-03 19:42:16 async_llm_engine.py:54]   File "/usr/lib/python3.10/concurrent/futures/thread.py", line 58, in run
ERROR 07-03 19:42:16 async_llm_engine.py:54]     result = self.fn(*self.args, **self.kwargs)
ERROR 07-03 19:42:16 async_llm_engine.py:54]   File "/usr/local/lib/python3.10/dist-packages/vllm/worker/worker_base.py", line 348, in execute_method
ERROR 07-03 19:42:16 async_llm_engine.py:54]     raise e
ERROR 07-03 19:42:16 async_llm_engine.py:54]   File "/usr/local/lib/python3.10/dist-packages/vllm/worker/worker_base.py", line 339, in execute_method
ERROR 07-03 19:42:16 async_llm_engine.py:54]     return executor(*args, **kwargs)
ERROR 07-03 19:42:16 async_llm_engine.py:54]   File "/usr/local/lib/python3.10/dist-packages/vllm/worker/worker_base.py", line 270, in execute_model
ERROR 07-03 19:42:16 async_llm_engine.py:54]     output = self.model_runner.execute_model(
ERROR 07-03 19:42:16 async_llm_engine.py:54]   File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 115, in decorate_context
ERROR 07-03 19:42:16 async_llm_engine.py:54]     return func(*args, **kwargs)
ERROR 07-03 19:42:16 async_llm_engine.py:54]   File "/usr/local/lib/python3.10/dist-packages/vllm/worker/model_runner.py", line 1276, in execute_model
ERROR 07-03 19:42:16 async_llm_engine.py:54]     hidden_or_intermediate_states = model_executable(
ERROR 07-03 19:42:16 async_llm_engine.py:54]   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
ERROR 07-03 19:42:16 async_llm_engine.py:54]     return self._call_impl(*args, **kwargs)
ERROR 07-03 19:42:16 async_llm_engine.py:54]   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1541, in _call_impl
ERROR 07-03 19:42:16 async_llm_engine.py:54]     return forward_call(*args, **kwargs)
ERROR 07-03 19:42:16 async_llm_engine.py:54]   File "/usr/local/lib/python3.10/dist-packages/vllm/model_executor/models/llama.py", line 400, in forward
ERROR 07-03 19:42:16 async_llm_engine.py:54]     model_output = self.model(input_ids, positions, kv_caches,
ERROR 07-03 19:42:16 async_llm_engine.py:54]   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
ERROR 07-03 19:42:16 async_llm_engine.py:54]     return self._call_impl(*args, **kwargs)
ERROR 07-03 19:42:16 async_llm_engine.py:54]   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1541, in _call_impl
ERROR 07-03 19:42:16 async_llm_engine.py:54]     return forward_call(*args, **kwargs)
ERROR 07-03 19:42:16 async_llm_engine.py:54]   File "/usr/local/lib/python3.10/dist-packages/vllm/model_executor/models/llama.py", line 297, in forward
ERROR 07-03 19:42:16 async_llm_engine.py:54]     hidden_states = self.get_input_embeddings(input_ids)
ERROR 07-03 19:42:16 async_llm_engine.py:54]   File "/usr/local/lib/python3.10/dist-packages/vllm/model_executor/models/llama.py", line 282, in get_input_embeddings
ERROR 07-03 19:42:16 async_llm_engine.py:54]     return self.embed_tokens(input_ids)
ERROR 07-03 19:42:16 async_llm_engine.py:54]   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
ERROR 07-03 19:42:16 async_llm_engine.py:54]     return self._call_impl(*args, **kwargs)
ERROR 07-03 19:42:16 async_llm_engine.py:54]   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1541, in _call_impl
ERROR 07-03 19:42:16 async_llm_engine.py:54]     return forward_call(*args, **kwargs)
ERROR 07-03 19:42:16 async_llm_engine.py:54]   File "/usr/local/lib/python3.10/dist-packages/vllm/prompt_adapter/layers.py", line 80, in forward
ERROR 07-03 19:42:16 async_llm_engine.py:54]     hidden_states[valid_mask] = gathered_embeddings
ERROR 07-03 19:42:16 async_llm_engine.py:54] IndexError: The shape of the mask [128] at index 0 does not match the shape of the indexed tensor [1, 5120] at index 0
```

SwapnilDreams100 commented:
Hmmm, @g-eoj can you share your test code?

Commit:
updating tests

Yard1 commented on tests/prompt_adapter/test_bloom.py:
```diff
@@ -0,0 +1,40 @@
+    return generated_texts
+
+
+def test_twitter_prompt_adapter():
+    llm = vllm.LLM(MODEL_PATH, enable_prompt_adapter=True, 
+                   max_prompt_adapter_token= 8)
+
+    expected_output = ['complaint', 'no complaint']
+
+    assert do_sample(llm, "twitter_pa", pa_id=1) == expected_output
```
this test fails if we do `LLM(enforce_eager=True)` (related to cuda graphs)

SwapnilDreams100 replied:
this has been resolved by the is_prompt check in the `model_runner`

g-eoj commented:
For https://github.com/vllm-project/vllm/pull/4645#issuecomment-2207110714, removing the enforce eager flag causes the error to go away.

Yard1 commented:
Yeah the fact it is not working with `enforce_eager=True` (which means no CUDA graphs) shows that there is a serious issue. We should test both cases.

Commit:
fix eager issue

SwapnilDreams100 commented:
Hey @g-eoj and @Yard1, the eager issue has been fixed, it was unrelated to the forward graph code.

SwapnilDreams100 committed:
Merge branch 'main' into main

Commit:
formatting

Commit:
formatting

Commit:
replacing numel w ndim for LoRA consistency

Yard1 commented on tests/prompt_adapter/test_bloom.py:
```diff
@@ -0,0 +1,42 @@
+        print(f"Prompt: {prompt!r}, Generated text: {generated_text!r}")
+    return generated_texts
+
+
+def test_twitter_prompt_adapter():
+    llm = vllm.LLM(MODEL_PATH,
+                   enforce_eager=True,
+                   enable_prompt_adapter=True,
+                   max_prompt_adapter_token=8)
```
```suggestion

@pytest.mark.parametrize("enforce_eager", [True, False])
def test_twitter_prompt_adapter(enforce_eager: bool):
    llm = vllm.LLM(MODEL_PATH,
                   enforce_eager=enforce_eager,
                   enable_prompt_adapter=True,
                   max_prompt_adapter_token=8)
```

Yard1 reviewed (commented):
@SwapnilDreams100 Actually I just realized this is only applied during prefill. In this case we do not need to do anything special for CUDA graph support. The current code looks fine and it should be decently performant. We should just add a comment to say that this will not work with CUDA graphs.

Yard1 commented on vllm/prompt_adapter/models.py:
```diff
@@ -0,0 +1,348 @@
+        for idx, id_ in enumerate(prompt_adapter_index_to_id)
+        if id_ is not None
+    }
+    pa_indices = torch.tensor([
+        id_to_index.get(id_, -1) if id_ > 0 else -1
+        for id_ in mapping.index_mapping
+    ])
+
+    pa_embedding_mapping = convert_to_embedding_indices(pa_indices)
```
```suggestion
    pa_indices = ([
        id_to_index.get(id_, -1) if id_ > 0 else -1
        for id_ in mapping.index_mapping
    ])

    pa_embedding_mapping = convert_to_embedding_indices(pa_indices)
    pa_indices = torch.tensor(pa_indices)
```

iterating over the list will be noticeably faster

SwapnilDreams100 committed:
Update tests/prompt_adapter/test_bloom.py

Co-authored-by: Antoni Baum <antoni.baum@protonmail.com>

SwapnilDreams100 committed:
Update vllm/prompt_adapter/models.py

Co-authored-by: Antoni Baum <antoni.baum@protonmail.com>

Commit:
formatting

Commit:
formatting

Yard1 reviewed (approved):
Approving pending CI passing

Commit:
formatting

Commit:
docs update

SwapnilDreams100 committed:
Merge pull request #2 from g-eoj/openai-entrypoint

Add openai entrypoint

Commit:
formatting

Commit:
formatting

Commit:
quick openapi fix

Commit:
formatting

Commit:
formatting

g-eoj commented:
@SwapnilDreams100 first of all congrats! 

I'd like to add tests for the openai server, would it be best to wait for this to merge so I can open my own PR against main?

SwapnilDreams100 commented:
Sounds good @g-eoj, big thank you for your help!

SwapnilDreams100 commented:
Hey @Yard1 are we good to merge?

njhill commented:
Thanks for this epic effort @SwapnilDreams100!! And big thanks to @Yard1 for the many detailed reviews.

I'll merge it before any new conflicts pop up!

SwapnilDreams100 commented:
Big thank you to @Yard1 for your guidance on this, this was a great learning experience!
